clippy_lints/src/derive.rs

```diff
@@ -156,11 +156,44 @@ declare_clippy_lint! {
     "deriving `serde::Deserialize` on a type that has methods using `unsafe`"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for types that derive `PartialEq` and could implement `Eq`.
+    ///
+    /// ### Why is this bad?
+    /// If a type `T` derives `PartialEq` and all of its members implement `Eq`,
+    /// then `T` can always implement `Eq`. Implementing `Eq` allows `T` to be used
+    /// in APIs that require `Eq` types. It also allows structs containing `T` to derive
+    /// `Eq` themselves.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[derive(PartialEq)]
+    /// struct Foo {
+    ///     i_am_eq: i32,
+    ///     i_am_eq_too: Vec<String>,
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// #[derive(PartialEq, Eq)]
+    /// struct Foo {
+    ///     i_am_eq: i32,
+    ///     i_am_eq_too: Vec<String>,
+    /// }
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub DERIVE_PARTIAL_EQ_WITHOUT_EQ,
+    nursery,
```

Comment:
Wasn't sure about the right place to put this.

Reply:
@rust-lang/clippy @dswij @Alexendoo @Jarcho Which category do you think is appropriate? I think `suspicious` looks good, but the original code isn't wrong.

Reply:
So this might be me being silly, but when I see a type that implements `PartialEq` but not `Eq`, I kind of read that as an assertion that "This is _not_ an equivalence relation, it's only a partial equivalence relation". In that sense I see it as suspicious to not implement `Eq` when you do have an equivalence relation, but that's pretty pedantic. I think you're correct in saying that this is primarily a question of adhering to API guidelines.

Reply:
The only reason not to implement `Eq` when it could be derived, other that laziness, is when there's the potential to extend the type in a way which could not implement `Eq`. This is not most cases, but it is a possibility. This possibility is, to me, enough of a reason to not categorize it as `suspicious`.

If not deriving `Eq` would cause the type to be misused, then `suspicious` would be a perfectly reasonable category. I think all our `suspicious` lints basically point out things which are easy to miswrite, or easy to misread in such a way that it creates a bug, but they are still reasonablish things to type. I haven't actually checked through all the `suspicious` lints, so that might not actually describe all `suspicious` lints.

Reply:
Completely agree with @Jarcho on this one. `sus` lints are for code that is "most likely wrong or useless". Not implementing `Eq` is not wrong, at least there is no way for Clippy to tell if it is intentional (possibly expanding the struct in the future). With that the API concern is the one left and for those, `style` is the best group.

Reply:
Thanks for the comments!